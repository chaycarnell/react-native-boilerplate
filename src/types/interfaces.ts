import { ScreenNames } from './enums';

export interface TranslationsMapping {
  [ScreenNames.DASHBOARD]: {
    description: string;
  };
  [ScreenNames.SECONDARY]: {
    description: string;
  };
}

export interface Translation {
  [translation: string]: TranslationsMapping;
}
