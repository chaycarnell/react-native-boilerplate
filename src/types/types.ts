import {ScreenNames} from './enums';

export type RootStackParamList = {
  [ScreenNames.DASHBOARD]: {};
  [ScreenNames.SECONDARY]: {};
};

export type ScreenName = `${ScreenNames}`;
