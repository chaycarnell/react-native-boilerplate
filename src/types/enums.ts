export enum ScreenNames {
  DASHBOARD = 'Dashboard',
  SECONDARY = 'Secondary',
}
