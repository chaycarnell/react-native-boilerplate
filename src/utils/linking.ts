import { Linking } from 'react-native';

// Handles opening a link
export const openLink = (link: string) => Linking.openURL(link);

// Function to handle linking actions of application
// Define linking actions here
export const handleLinking = async (event: any) =>
  console.log('Linking event: ', event);

// Handle universal linking where app was closed and launched from an associated URL
export const checkUniversalLinkState = () =>
  Linking.getInitialURL()
    .then((url: any) => {
      if (url) handleLinking(url);
    })
    .catch((e: any) => console.log('Linking error: ', e));

// Handles deep linking or universal links where app is woken up from background
export const applyDeepLinkListener = () =>
  Linking.addEventListener('url', handleLinking);
