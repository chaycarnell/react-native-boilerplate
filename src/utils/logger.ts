import packageJson from '../../package.json';

type LoggingFunction = (...args: Array<unknown>) => void;

type Logger = {
  debug: LoggingFunction;
  error: LoggingFunction;
  info: LoggingFunction;
  log: LoggingFunction;
  trace: LoggingFunction;
  warn: LoggingFunction;
};

const packageInfo = () => `${packageJson.name} (${packageJson.version})`;

export const logger: Logger = {
  debug: (...args) => console.debug(packageInfo(), ...args),
  error: (...args) => console.error(packageInfo(), ...args),
  info: (...args) => console.info(packageInfo(), ...args),
  log: (...args) => console.log(packageInfo(), ...args),
  trace: (...args) => console.trace(packageInfo(), ...args),
  warn: (...args) => console.warn(packageInfo(), ...args),
};
