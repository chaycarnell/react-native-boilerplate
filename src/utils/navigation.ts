import { BackHandler } from 'react-native';
import { createNavigationContainerRef } from '@react-navigation/native';

// Initialise a root navigation ref
export const navigator = createNavigationContainerRef();

// Custom navigate handler allowing stack to be reset in props
export const navigate = (
  screenName: string,
  params: any,
  resetStack: boolean | undefined = false,
) => {
  if (resetStack)
    return navigator.reset({
      index: 0,
      routes: [{ name: screenName, params }],
    });
  // @ts-expect-error
  return navigator.navigate(screenName, params);
};

// Returns a boolean based on if the user can go back further than the current screen
export const canGoBack = () => navigator.canGoBack();

// Prevent hardware back if back options not available
// This prevents the app closing in Android
const handleHardwareBack = () => !canGoBack();

// Handles applying hardware back action event listener
export const applyBackHandleListener = () =>
  BackHandler.addEventListener('hardwareBackPress', handleHardwareBack);

// Handles removing hardware back action event listener
export const removeBackHandleListener = () =>
  BackHandler.removeEventListener('hardwareBackPress', handleHardwareBack);

export type RootStackParamList = {
  Auth: {};
  Dashboard: {};
};
