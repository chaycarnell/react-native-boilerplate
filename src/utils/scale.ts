import {Dimensions} from 'react-native';

const {width, height} = Dimensions.get('window');
const [shortDimension, longDimension] =
  width < height ? [width, height] : [height, width];

// Default guideline sizes are based on standard ~5" screen mobile device
// Suggested to set this based on the design base dimensions or device model
// Note, this approach generally works well with apps fixed to a static screen orientation, i.e. portrait
const guidelineBaseWidth = 375;
const guidelineBaseHeight = 812;

export const scale = (size: number) =>
  (shortDimension / guidelineBaseWidth) * size;
export const verticalScale = (size: number) =>
  (longDimension / guidelineBaseHeight) * size;
export const moderateScale = (size: number, factor: number = 0.5) =>
  size + (scale(size) - size) * factor;

export const s = scale;
export const vs = verticalScale;
export const ms = moderateScale;
