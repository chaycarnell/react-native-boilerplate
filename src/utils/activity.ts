import { AppState, AppStateStatus } from 'react-native';

import { logger } from './logger';

// Handles listener actions for device state changes
// Define state change actions here
const handleStateChange = (state: AppStateStatus) => {
  if (state === 'active') {
    logger.info('Perform active state actions here');
  }
  if (state === 'background') {
    logger.info('Perform background state actions here');
  }
};

// Applies listener actions to the device state, i.e. 'active', 'background', 'inactive'
// This is useful for example if you want to fetch new data on app reopen from being in background mode
export const applyStateListeners = () =>
  AppState.addEventListener('change', handleStateChange);
