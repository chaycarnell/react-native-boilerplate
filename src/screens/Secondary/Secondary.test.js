import {cleanup, fireEvent, render} from '@testing-library/react-native';
import {ScreenNames} from '@src/types';
import * as NavModule from '@src/utils/navigation';
import React from 'react';

import AuthScreen from './Secondary';

afterEach(cleanup);

describe('Secondary screen', () => {
  describe('given default props', () => {
    let navSpy;
    let renderedComponent;

    beforeEach(() => {
      navSpy = jest.spyOn(NavModule, 'navigate').mockImplementation(jest.fn);
      renderedComponent = render(<AuthScreen />);
    });

    afterAll(() => {
      navSpy.mockRestore();
    });

    it('should render correctly', () => {
      const {toJSON} = renderedComponent;
      expect(toJSON()).toMatchSnapshot();
    });

    it('should navigate to dashboard page when touchable highlight is selected', () => {
      const {getByTestId} = renderedComponent;
      fireEvent.press(getByTestId('SecondaryNavigationButton'));
      expect(NavModule.navigate).toHaveBeenCalledWith(
        ScreenNames.DASHBOARD,
        {},
      );
    });
  });
});
