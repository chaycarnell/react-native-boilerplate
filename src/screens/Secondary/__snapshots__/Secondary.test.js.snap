// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secondary screen given default props should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="SecondaryNavigationButton"
  >
    <View
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#0f62fe",
          "display": "flex",
          "height": 667,
          "justifyContent": "center",
          "opacity": 1,
          "width": 750,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 24,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      />
    </View>
  </View>
</View>
`;
